### PR TITLE
Add share attributes + prevent download permission

### LINKED ISSUE #16413: Download Permission restriction like Hide Download option for internal (user/group) shares
hide download

Hide Download option is already in Share link options in Nextcloud. But the option isn't available if you share with another user / group. The idea is to create a download permission like a hide download option in share link.

I think this feature very needed in real world case especially for strict permission.

### PR DESCRIPTION
**Pull request by PVince81**
Close https://github.com/nextcloud/server/issues/16413

- brings in share attributes column in oc_share
- store download permission in extended share attributes
- enforces download restriction

+ [x] backend
+ [x] frontend, add "hide download" permission
+ [ ] ~~extension points for secure view ?~~ not for now
+ [ ] BUG: share view-only file with link, open link then download => this should be blocked on backend level and in UI
+ [ ] BUG: share view-only folder with link, open link then download any entry => this should be blocked on backend level and in UI
+ [x] BUG: reshare with public link should automatically tick "hide download"
+ [x] BUG: reshare with public link should not allow to tick "hide download" 
+ [ ] BUG: viewer app still opens (optional)
+ [x] TEST: try increasing attribute permissions in reshare create
+ [x] TEST: try increasing attribute permissions in reshare update
+ [ ] TODO: make sure versions endpoint also doesn't allow download past versions

### CODE REVIEW

**Comment by PVince81** [comment]
regardless how many times I run `composer run cs:fix` locally it doesn't fix anything :rage:

**Comment by PVince81** [comment]
for testing, payload to set with curl on a share (here with id 2):
```
curl -u admin:admin 'https://vvortex.local/ocs/v2.php/apps/files_sharing/api/v1/shares/2' \
  -X 'PUT' \
  -H 'OCS-APIRequest: true' \
-H 'Accept: application/json, text/plain, */*' \
  -H 'Content-Type: application/json' \
  --data-raw '{"attributes":[{"scope": "permission", "key": "download", "enabled": "false"}]}'
```

**Comment by PVince81** [comment]
- [ ] consider renaming "enabled" to "value" as we likely want more than a boolean there for flexibility ? (also change the method interfaces then)

**Comment by PVince81** [comment]
sqlite only not reproducible locally :open_mouth: 
```
1) Test\Share20\DefaultShareProviderTest::testGetShareByIdUserShare
Invalid argument supplied for foreach()

/drone/src/lib/private/Share20/DefaultShareProvider.php:1563
/drone/src/lib/private/Share20/DefaultShareProvider.php:1082
/drone/src/lib/private/Share20/DefaultShareProvider.php:814
/drone/src/tests/lib/Share20/DefaultShareProviderTest.php:221
```

**Comment by PVince81** [comment]
rebased for CI, hopefully everything green after that!

**Comment by PVince81** [comment]
- [ ] simplify permissions in DB ?
   - ex: `["permissions","download",false]` or even: `{'permissions': {'download': false}}`

**Comment by PVince81** [comment]
I've added a "Allow download" checkbox now in the frontend and also the logic to read and save it.
Additionally, the share attributes are now available in a "nc:share-attributes" DAV property for clients and is also used by the web UI.

More work needed to:
- [ ] make sure resharing cannot increase download perms
- [ ] make sure public link page doesn't show download action when resharing a received share that has no download perm

- [ ] see if possible to have JS tests for the new bits (at first look it seems we don't have any)

File: apps/files_sharing/lib/AppInfo/Application.php [hunk 1]
```diff
@@ -139,18 +143,76 @@ public function registerEventsScripts(IEventDispatcher $dispatcher, EventDispatc
...
+	public function registerDownloadEvents(
+		IEventDispatcher $dispatcher,
+		?IUserSession $userSession,
+		IRootFolder $rootFolder
+	) {
+
+		$dispatcher->addListener(
+			'file.beforeGetDirect',
+			function (GenericEvent $event) use ($userSession, $rootFolder) {
+				$pathsToCheck[] = $event->getArgument('path');
```
Comment (PVince81) [inline apps/files_sharing/lib/AppInfo/Application.php:167]: move the square brackets to the right side

File: apps/files_sharing/lib/Controller/ShareAPIController.php [hunk 2]
```diff
@@ -456,7 +462,8 @@ public function createShare(
 		string $sendPasswordByTalk = null,
 		string $expireDate = '',
 		string $note = '',
-		string $label = ''
+		string $label = '',
+		string $attributes = null
```
Comment (PVince81) [inline apps/files_sharing/lib/Controller/ShareAPIController.php:466]: `?string`

File: apps/files_sharing/lib/Controller/ShareAPIController.php [hunk 3]
```diff
@@ -1051,7 +1063,8 @@ public function updateShare(
 		string $expireDate = null,
 		string $note = null,
 		string $label = null,
-		string $hideDownload = null
+		string $hideDownload = null,
+		string $attributes = null
```
Comment (PVince81) [inline apps/files_sharing/lib/Controller/ShareAPIController.php:1067]: `?string`

File: apps/files_sharing/lib/ViewOnly.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+	public function __construct(Folder $userFolder) {
+		$this->userFolder = $userFolder;
+	}
+
+	/**
+	 * @param string[] $pathsToCheck
+	 * @return bool
+	 */
+	public function check($pathsToCheck) {
```
Comment (PVince81) [inline apps/files_sharing/lib/ViewOnly.php:45]: add strict type

File: apps/files_sharing/src/components/SharingEntry.vue [hunk 5]
```diff
@@ -390,6 +426,10 @@ export default {
 				| (isReshareChecked ? this.permissionsShare : 0)
 
 			this.share.permissions = permissions
+			if (this.share.hasDownloadPermission !== isDownloadChecked) {
+				this.share.hasDownloadPermission = isDownloadChecked
+				this.queueUpdate('attributes')
+			}
```
Comment (PVince81) [inline apps/files_sharing/src/components/SharingEntry.vue:432]: optional: check if we can also skip "permissions"

File: lib/private/legacy/OC_Files.php [hunk 6]
```diff
@@ -167,6 +169,14 @@ public static function get($dir, $files, $params = null) {
 				}
 			}
 
+			//Dispatch an event to see if any apps have problem with download
+			$event = new GenericEvent(null, ['dir' => $dir, 'files' => $files, 'run' => true]);
```
Comment (PVince81) [inline lib/private/legacy/OC_Files.php:173]: use custom class

File: lib/public/Share/IAttributes.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+namespace OCP\Share;
+
+/**
+ * Interface IAttributes
```
Comment (PVince81) [inline lib/public/Share/IAttributes.php:24]: add message for "use only"

File: lib/public/Share/IShare.php [hunk 8]
```diff
@@ -314,6 +314,31 @@ public function setPermissions($permissions);
 	 */
 	public function getPermissions();
 
+	/**
+	 * Create share attributes object
+	 *
+	 * @since 25.0.0
+	 * @return IAttributes
+	 */
+	public function newAttributes();
```
Comment (PVince81) [inline lib/public/Share/IShare.php:323]: add return type

File: lib/public/Share/IShare.php [hunk 9]
```diff
@@ -314,6 +314,31 @@ public function setPermissions($permissions);
...
+
+	/**
+	 * Set share attributes
+	 *
+	 * @param ?IAttributes $attributes
+	 * @since 25.0.0
+	 * @return IShare The modified object
+	 */
+	public function setAttributes(?IAttributes $attributes);
```
Comment (PVince81) [inline lib/public/Share/IShare.php:332]: add "use only" on the interface

File: apps/dav/lib/DAV/ViewOnlyPlugin.php [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	/** @var Server $server */
+	private $server;
+
+	/** @var ILogger $logger */
+	private $logger;
+
+	/**
+	 * @param ILogger $logger
+	 */
+	public function __construct(ILogger $logger) {
```
Comment (juliushaertl) [inline apps/dav/lib/DAV/ViewOnlyPlugin.php:49]: Switch to LoggerInterface?

File: lib/private/legacy/OC_Files.php [hunk 11]
```diff
@@ -322,6 +338,19 @@ private static function getSingleFile($view, $dir, $name, $params) {
 			$rangeArray = self::parseHttpRangeHeader(substr($params['range'], 6), $fileSize);
 		}
 
+		$dispatcher = \OC::$server->query(IEventDispatcher::class);
+		$event = new GenericEvent(null, ['path' => $filename]);
+		$dispatcher->dispatch('file.beforeGetDirect', $event);
```
Comment (juliushaertl) [inline lib/private/legacy/OC_Files.php:343]: Could be a separate event class. There is already https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/apps/dav/lib/Events/BeforeFileDirectDownloadedEvent.php#L34 but that covers a different endpoint

**Comment by juliushaertl** [comment]
Just had a first look at the code so far, not tested yet

- [ ] Probably needs some extra handling for the direct download endpoint that mobile clients may use (ref https://github.com/nextcloud/server/pull/9178)

**Comment by PVince81** [comment]
@juliushaertl any comments on the data structure for the permissions ? I have mixed feelings there as it could probably be simplified, at least outside the database.

Is this something that apps like richdocuments could easily reuse ?

**Comment by PVince81** [comment]
I've sent a separate PR just for the column migration so that we can have it in earlier: https://github.com/nextcloud/server/pull/32746

File: apps/files_sharing/src/models/Share.js [hunk 12]
```diff
@@ -43,6 +43,14 @@ export default class Share {
 		ocsData.hide_download = !!ocsData.hide_download
 		ocsData.mail_send = !!ocsData.mail_send
 
+		if (ocsData.attributes) {
+			try {
+				ocsData.attributes = JSON.parse(ocsData.attributes)
+			} catch (e) {
+				console.warn('Could not parse share attributes returned by server: "' + ocsData.attributes + '"')
+			}
+		}
```
Comment (juliushaertl) [inline apps/files_sharing/src/models/Share.js:52]: As the later parts assume that attributes is an array:

```suggestion
		} else {
			ocsData.attributes = []
		}
```

This makes sure that changing the attributes works properly in the frontend if none are present yet

File: apps/files_sharing/src/components/SharingEntry.vue (hunk 5 above)
Comment (juliushaertl) [reply apps/files_sharing/src/components/SharingEntry.vue:432]: Would be good, seems with the current approach there are always two PUT requests. queueUpdate should allow to pass multiple properties to update: https://github.com/nextcloud/server/blob/e1a36bedb40f4ef4dfecc62cb7fac30256bba26f/apps/files_sharing/src/mixins/SharesMixin.js#L220

**Comment by PVince81** [comment]
- [ ] BUG: share view-only file with link, open link then download => this should be blocked
- [ ] BUG: reshare with public link should automatically tick "hide download"
- [ ] BUG: viewer app still opens

**Comment by PVince81** [comment]
- [ ] TODO: make sure versions endpoint also doesn't allow download past versions

**Comment by PVince81** [comment]
I've fixed the DirectDownloadController to also fire the event, now it will return a 403 if a download is denied.

**Comment by PVince81** [comment]
I managed to map the download permission to "hide_download" column at creation and update time

the UI still needs to be adjusted to prevent changing that permission

File: apps/files_sharing/lib/Controller/ShareAPIController.php [hunk 13]
```diff
@@ -542,6 +550,14 @@ public function createShare(
 				throw new OCSNotFoundException($this->l->t('Public link sharing is disabled by the administrator'));
 			}
 
+			if ($share->getNode()->getStorage()->instanceOfStorage(SharedStorage::class)) {
```
Comment (CarlSchwan) [inline apps/files_sharing/lib/Controller/ShareAPIController.php:553]: Shouldn't this apply to all the share types? if one user share me a file, I shouldn't be able to share it with another user with more permissions

Comment (PVince81) [reply apps/files_sharing/lib/Controller/ShareAPIController.php:553]: yes. IIRC this might already be enforced on a different level

**Comment by PVince81** [comment]
:+1: code changes look fine

since I'm the PR submitter I can't approve it

@juliushaertl can you approve also ?

**Comment by PVince81** [comment]
@CarlSchwan please squash. and while you're at it, one of the new event files has warnings on the header so probably you can fix them in that go

**Review by miaulalala — APPROVED** [review]

**Comment by PVince81** [comment]
some more failures: https://drone.nextcloud.com/nextcloud/server/21540/9/4
```
1) OCA\DAV\Tests\Unit\DAV\Controller\DirectControllerTest::testGetUrlNonExistingFileId
Error: Object of class Mock_IURLGenerator_d72bc28a could not be converted to string

/drone/src/apps/dav/tests/unit/Controller/DirectControllerTest.php:87
```

reproducible locally

**Comment by PVince81** [comment]
please merge once CI passes!

**Review by CarlSchwan — APPROVED** [review]
ci issue unrelated

**Comment by elhananjair** [comment]
Hello, @PVince81 
I was waiting for this commit for so long, I tested it recently, and didn't work well with PDF files 😢. It works fine with other files I tested such as .docx .md. 

Screenshot from 2022-08-17 20-23-40
Screenshot from 2022-08-17 20-50-31

**Comment by PVince81** [comment]
@elhananjair yes, the different apps also need to be adjusted to properly recognize the permission and act accordingly.
for apps that don't support view only, they should simply not view the document at all

**Comment by juliushaertl** [comment]
I filed https://github.com/nextcloud/files_pdfviewer/issues/649 about this

**Comment by iMoshi** [comment]
So, I've finally tried the feature, and unchecked "Allow download" when sharing a folder with an internal user, and the user would get "Access forbidden" with basically any document including .pdf.

Is there something I'm missing inside the Administration Settings?

Also, can we have the option to "Allow download" when sharing a single file with an internal user?

**Pull request merged**